letsencrypt-nginx/letsencrypt_nginx/nginxparser.py

```diff
@@ -1,56 +1,241 @@
-"""Very low-level nginx config parser based on pyparsing."""
-import string
+"""Very low-level hand rolled nginx config parser, using recursive descent to
+retain all parsed information, so that the file can be reconstituted."""
 
-from pyparsing import (
-    Literal, White, Word, alphanums, CharsNotIn, Forward, Group,
-    Optional, OneOrMore, Regex, ZeroOrMore)
-from pyparsing import stringEnd
-from pyparsing import restOfLine
+import StringIO
 
+class ParseException(Exception):
+    """A parsing exception happened."""
+    def __init__(self, value):
+        self.value = value
+        super(ParseException, self).__init__()
+
+    def __str__(self):
+        return repr(self.value)
 
 class RawNginxParser(object):
     # pylint: disable=expression-not-assigned
-    """A class that parses nginx configuration with pyparsing."""
-
-    # constants
-    left_bracket = Literal("{").suppress()
-    right_bracket = Literal("}").suppress()
-    semicolon = Literal(";").suppress()
-    space = White().suppress()
-    key = Word(alphanums + "_/")
-    # Matches anything that is not a special character AND any chars in single
-    # or double quotes
-    value = Regex(r"((\".*\")?(\'.*\')?[^\{\};,]?)+")
-    location = CharsNotIn("{};," + string.whitespace)
-    # modifier for location uri [ = | ~ | ~* | ^~ ]
-    modifier = Literal("=") | Literal("~*") | Literal("~") | Literal("^~")
-
-    # rules
-    comment = Literal('#') + restOfLine()
-    assignment = (key + Optional(space + value, default=None) + semicolon)
-    location_statement = Optional(space + modifier) + Optional(space + location)
-    if_statement = Literal("if") + space + Regex(r"\(.+\)") + space
-    block = Forward()
-
-    block << Group(
-        (Group(key + location_statement) ^ Group(if_statement)) +
-        left_bracket +
-        Group(ZeroOrMore(Group(comment | assignment) | block)) +
-        right_bracket)
-
-    script = OneOrMore(Group(comment | assignment) ^ block) + stringEnd
+    """A class that parses nginx configuration."""
 
     def __init__(self, source):
         self.source = source
 
-    def parse(self):
-        """Returns the parsed tree."""
-        return self.script.parseString(self.source)
-
     def as_list(self):
         """Returns the parsed tree as a list."""
-        return self.parse().asList()
+        return self.parse()
 
+    def parse(self):
+        """Returns the parsed tree."""
+        return RawNginxParser.Internal(self.source).parse_file()
+
+    class Internal(object):
```

Comment:
Why split this off as a separate class? It's actually increasing the amount of state, since both the external and the internal class need to hang onto `source`/`src`. There's no additional state in the outer class.

In fact, the outer class should probably go away altogther. (A good rule of thumb is that a class with only a single public method should just be a function. RawNginxParser has 2, but one just calls the other. (Why?)) Why force callers to construct a class and call a method on it when there's nothing else useful you can do with the class? We could simplify the API to something like…

```python
foo = parse_nginx('some config code')
```

If you want to design for subclassing, that could even be a class method.